File: algorithms/java/QuickSort.java
public class QuickSort
{
	private static int partition(int[] array, int start, int end)
	{
		int pivot = array[start];
		int median = start;

		for(int i = start + 1; i <= end; i++)
		{
			if(array[i] < pivot)
			{
				median++;
				int a = array[i];
				int b = array[median];
				array[i] = b;
				array[median] = a;
			}
		}

		int x = array[start];
		int y = array[median];
		array[start] = y;
		array[median] = x;

		return median;
	}

	public static void sort(int[] array, int start, int end)
	{
		if(start < end)
		{
			int pivot = partition(array, start, end);

			sort(array, start, pivot - 1);
			sort(array, pivot + 1, end);
		}
	}
}

File: algorithms/typescript/quick-sort.ts
function partition(array: number[], start: number, end: number)
{
	let pivot = array[start];
	let median = start;

	for(let i = start + 1; i <= end; i++)
	{
		if(array[i] < pivot)
		{
			median++;
			let a = array[i];
			let b = array[median];
			array[i] = b;
			array[median] = a;
		}
	}

	let x = array[start];
	let y = array[median];
	array[start] = y;
	array[median] = x;

	return median;
}

function quickSort(array: number[], start: number, end: number)
{
	if(start < end)
	{
		let pivot = partition(array, start, end);

		quickSort(array, start, pivot - 1);
		quickSort(array, pivot + 1, end);
	}

	return array;
}